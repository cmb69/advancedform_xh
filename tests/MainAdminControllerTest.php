<?php

namespace Advancedform;

use ApprovalTests\Approvals;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\MockObject\Stub;
use PHPUnit\Framework\TestCase;
use Plib\CsrfProtector;
use Plib\FakeRequest;
use Plib\Random;
use Plib\View;
use XH\Pages;

class MainAdminControllerTest extends TestCase
{
    /** @var FormGateway */
    private $formGateway;

    /** @var CsrfProtector&Stub */
    private $csrfProtector;

    /** @var Pages&Stub */
    private $pages;

    /** @var Random&Stub */
    private $random;

    public function setUp(): void
    {
        global $tx;
        $tx = XH_includeVar("../../cmsimple/languages/en.php", "tx");
        $this->setUpFormGateway();
        $this->csrfProtector = $this->createStub(CsrfProtector::class);
        $this->csrfProtector->method("token")->willReturn("0123456789ABCDEF");
        $this->pages = $this->createStub(Pages::class);
        $this->pages->method("linkList")->willReturn([]);
        $this->random = $this->createStub(Random::class);
        $this->random->method("bytes")->willReturn("0123456789ABCDE");
    }

    private function setUpFormGateway(): void
    {
        vfsStream::setup("root");
        copy("./data/forms.json", vfsStream::url("root/forms.json"));
        $this->formGateway = new FormGateway(vfsStream::url("root/"));
        $contents = file_get_contents(vfsStream::url("root/forms.json"));
        $forms = json_decode($contents, true);
        $forms = ["Import" => $forms["Contact"], '%VERSION%' => Plugin::DB_VERSION];
        $contents = json_encode($forms, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        file_put_contents(vfsStream::url("root/Import.json"), $contents);
        mkdir(vfsStream::url("root/css"));
    }

    private function sut(): MainAdminController
    {
        return new MainAdminController(
            $this->formGateway,
            "./",
            XH_includeVar("./config/config.php", "plugin_cf")["advancedform"],
            XH_includeVar("./languages/en.php", "plugin_tx")["advancedform"],
            $this->csrfProtector,
            $this->pages,
            $this->random,
            new View("./templates/", XH_includeVar("./languages/en.php", "plugin_tx")["advancedform"])
        );
    }

    public function testRendersFormsOverview(): void
    {
        $response = $this->sut()(new FakeRequest());
        Approvals::verifyHtml($response->output());
    }

    public function testRendersFormEditor(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=edit&form=Contact",
        ]);
        $response = $this->sut()($request);
        Approvals::verifyHtml($response->output());
    }

    public function testCreatesNewForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = ["advancedform_token" => "0123456789ABCDEF"];
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=new",
        ]);
        $response = $this->sut()($request);
        $this->assertArrayHasKey("60OJ4CPK6KR3EE1P85146H25", $this->formGateway->findAll());
        $this->assertSame(
            "http://example.com/?advancedform&admin=plugin_main&action=edit&form=60OJ4CPK6KR3EE1P85146H25",
            $response->location()
        );
    }

    public function testSavingIsCsrfProtected(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = [
            "advancedform_token" => "0123456789ABCDEF",
        ];
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=save&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("nope", $response->output());
    }

    public function testSavingReportsMissingForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = [
            "advancedform_token" => "0123456789ABCDEF",
        ];
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=save&form=NoSuchForm",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("A form with the name 'NoSuchForm' does not exist!", $response->output());
    }

    public function testSavingReportsExistingForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = $this->formPost("Remko");
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=save&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("A form with this name already exists!", $response->output());
    }

    public function testSavingReportsFailureToSave(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = $this->formPost("Contact");
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=save&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("The forms database could not have been saved!", $response->output());
    }

    public function testSavesForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = $this->formPost("Saved");
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=save&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertArrayHasKey("Saved", $this->formGateway->findAll());
        $this->assertSame(
            "http://example.com/?advancedform&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testDeletingIsCsrfProtected(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = [
            "advancedform_token" => "0123456789ABCDEF",
        ];
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=delete&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("nope", $response->output());
    }

    public function testDeletingReportsMissingForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = [
            "advancedform_token" => "0123456789ABCDEF",
        ];
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=delete&form=NoSuchForm",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("A form with the name 'NoSuchForm' does not exist!", $response->output());
    }

    public function testDeletingReportsFailureToSave(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = $this->formPost("Contact");
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=delete&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("The forms database could not have been saved!", $response->output());
    }

    public function testDeletesForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = $this->formPost("Contact");
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=delete&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertArrayNotHasKey("Contact", $this->formGateway->findAll());
        $this->assertSame(
            "http://example.com/?advancedform&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testCopyingIsCsrfProtected(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = [
            "advancedform_token" => "0123456789ABCDEF",
        ];
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=copy&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("nope", $response->output());
    }

    public function testCopyingReportsMissingForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = [
            "advancedform_token" => "0123456789ABCDEF",
        ];
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=copy&form=NoSuchForm",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("A form with the name 'NoSuchForm' does not exist!", $response->output());
    }

    public function testCopyingReportsFailureToSave(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = $this->formPost("Contact");
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=copy&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("The forms database could not have been saved!", $response->output());
    }

    public function testCopiesForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $_POST = $this->formPost("Contact");
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=copy&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertArrayHasKey("60OJ4CPK6KR3EE1P85146H25", $this->formGateway->findAll());
        $this->assertSame(
            "http://example.com/?advancedform&admin=plugin_main&action=edit&form=60OJ4CPK6KR3EE1P85146H25",
            $response->location()
        );
    }

    public function testImportingIsCsrfProtected(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=import&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("nope", $response->output());
    }

    public function testImportingReportsExistingForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=import&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("A form with this name already exists!", $response->output());
    }

    public function testImportingReportsFailureToImport(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=import&form=New",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("'vfs://root/New.json' could not have been imported!", $response->output());
    }

    public function testImportingReportsFailureToSave(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=import&form=Import",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("The forms database could not have been saved!", $response->output());
    }

    public function testImportsForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=import&form=Import",
        ]);
        $response = $this->sut()($request);
        $this->assertArrayHasKey("Import", $this->formGateway->findAll());
        $this->assertSame(
            "http://example.com/?advancedform&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testExportingIsCsrfProtected(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=export&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("nope", $response->output());
    }

    public function testExportingReportsMissingForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=export&form=NoSuchForm",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("A form with the name 'NoSuchForm' does not exist!", $response->output());
    }

    public function testExportingReportsFailureToExport(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=export&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString(
            "'vfs://root/Contact.json' could not have been exported!",
            $response->output()
        );
    }

    public function testExportsForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=export&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertFileExists(vfsStream::url("root/Contact.json"));
        $this->assertSame(
            "http://example.com/?advancedform&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testCreatingFormTemplateIsCsrfProtected(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=template&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("nope", $response->output());
    }

    public function testCreatingFormTemplateReportsMissingForm(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=template&form=NoSuchForm",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("A form with the name 'NoSuchForm' does not exist!", $response->output());
    }

    public function testCreatingFormTemplateReportsFailureToSave(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=template&form=Contact",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString(
            "'vfs://root/Contact.tpl' could not be saved!",
            $response->output()
        );
    }

    public function testCreatesFormTemplate(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=template&form=Contact",
        ]);
        $response = $this->sut()($request);
        Approvals::verifyHtml(file_get_contents(vfsStream::url("root/Contact.tpl")));
        $this->assertSame(
            "http://example.com/?advancedform&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testCreatesFormTemplateCSS(): void
    {
        $_SERVER["REQUEST_METHOD"] = "POST";
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?advancedform&admin=plugin_main&action=template&form=Contact",
        ]);
        $response = $this->sut()($request);
        Approvals::verifyStringWithFileExtension(file_get_contents(vfsStream::url("root/css/Contact.css")), "css");
        $this->assertSame(
            "http://example.com/?advancedform&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    private function formPost(string $name): array
    {
        return [
            "advancedform_token" => "0123456789ABCDEF",
            "advfrm-name" => $name,
            "advfrm-title" => "",
            "advfrm-to_name" => "",
            "advfrm-to" => "",
            "advfrm-cc" => "",
            "advfrm-bcc" => "",
            "advfrm-thanks_page" => "",
            "advfrm-field" => [
                "",
            ],
            "advfrm-label" => [
                "",
            ],
            "advfrm-type" => [
                "",
            ],
            "advfrm-props" => [
                "",
            ],
            "advfrm-required" => [
                "",
            ],
        ];
    }
}
