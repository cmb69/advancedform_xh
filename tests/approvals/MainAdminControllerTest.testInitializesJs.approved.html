<meta name="advancedform.config" content='{"syscheck_title":"System check","syscheck_phpversion":"PHP version ≥ %s","syscheck_extension":"Extension \u0027%s\u0027 loaded","syscheck_xhversion":"CMSimple_XH version ≥ %s","syscheck_plugin":"Plugin \u0027%s\u0027 installed","syscheck_captcha_plugin":"CAPTCHA installed","syscheck_captcha_key":"CAPTCHA key configured","syscheck_writable":"Folder \u0027%s\u0027 writable","contact_form":"","menu_main":"Mail Forms","mail_subject":"%1$s on %2$s","mail_subject_confirmation":"%1$s from %2$s","captcha_explanation":"Please additionally enter the shown digits:","date_format":"m/d/Y","message_confirm_delete":"The form and all it\u0027s information will be deleted!","message_confirm_export":"A possibly existing %s.json file will be overwritten!","message_confirm_template":"A possibly existing %1$s.tpl and %1$s.css file will be overwritten!","message_sent_info":"Thank you for filling out the form. The following information has been sent:","message_required_field":"required fields","message_required_fields":"Please fill in all %s!","message_script_code":"Script code to insert on a page","message_import_form":"Name of the mail form to import","label_name":"Name","label_title":"Title","label_to_name":"To (name)","label_to":"To (e-mail)","label_cc":"CC","label_bcc":"BCC","label_captcha":"CAPTCHA","label_store":"Store data","label_thanks_page":"Thanks page","label_field":"Name","label_label":"Label","label_type":"Type","label_required":"Req.","label_properties":"%s Properties","label_size":"Size","label_cols":"Columns","label_maxlen":"max. length","label_rows":"Rows","label_default":"Default","label_value":"Value","label_ftypes":"File types","label_constraint":"Constraint","label_error_msg":"Error message","label_horizontal":"Horizontal","label_vertical":"Vertical","label_none":"NONE","field_text":"Text","field_from_name":"Sender (name)","field_from":"Sender (e-mail)","field_mail":"E-Mail","field_date":"Date","field_number":"Number","field_textarea":"Textarea","field_radio":"Radiobutton","field_select":"Selectbox","field_checkbox":"Checkbox","field_multi_select":"Multi-Selectbox","field_password":"Password","field_file":"File","field_hidden":"Hidden","field_output":"Output","field_custom":"Custom","tool_add":"Add","tool_delete":"Delete","tool_down":"Down","tool_up":"Up","tool_props":"Properties","tool_copy":"Copy","tool_clear_defaults":"Clear defaults","tool_export":"Export","tool_import":"Import","tool_template":"Create template","button_ok":"Ok","button_cancel":"Cancel","button_send":"Send mail","button_reset":"Reset form","error_mail":"The e-mail could not be sent!","error_missing_sender":"Missing sender e-mail address! Confirmation mail could not be sent!","error_missing_field":"Required field \u0027%s\u0027 is empty!","error_invalid_email":"Field \u0027%s\u0027 doesn\u0027t contain a valid e-mail address!","error_invalid_number":"Field \u0027%s\u0027 doesn\u0027t contain a valid number!","error_invalid_date":"Field \u0027%s\u0027 doesn\u0027t contain a valid date!","error_invalid_custom":"The content of field \u0027%s\u0027 is not valid!","error_upload_too_large":"The file attachment of field \u0027%s\u0027 is too large!","error_upload_illegal_ftype":"The file type \u0027%2$s\u0027 of field \u0027%1$s\u0027 can\u0027t be attached!","error_upload_general":"The file of field \u0027%s\u0027 can\u0027t be attached!","error_captcha_code":"Wrong CAPTCHA code!","error_form_exists":"A form with this name already exists!","error_form_missing":"A form with the name \u0027%s\u0027 does not exist!","error_missing_fields":"Please fill in the required fields!","error_invalid_field":"Field name is invalid!","error_default_too_long":"The default value is longer than the max. length!","error_invalid_property":"Invalid property value!","error_duplicate_fields":"Duplicate field names are not allowed!","error_too_many_senders":"At most one sender allowed!","error_export":"\u0027%s\u0027 could not have been exported!","error_import":"\u0027%s\u0027 could not have been imported!","error_save":"The forms database could not have been saved!","error_template":"\u0027%s\u0027 could not be saved!","error_captcha":"Could not load CAPTCHA!","error_csv":"Could not store form data!","log_success":"Email from <%s> sent.","log_error":"Email from <%s> not sent!"}'>
<script src="./plugins/advancedform/admin.min.js"></script>
